code/ship/ship.cpp

```diff
@@ -9072,46 +9127,54 @@ static void ship_auto_repair_frame(int shipnum, float frametime)
 	sip = &Ship_info[sp->ship_info_index];
 	objp = &Objects[sp->objnum];
 
-	if (sp->flags[Ship::Ship_Flags::Dying]) // do not repair if already dead 
+	// do not repair if already dead 
+	if (sp->flags[Ship::Ship_Flags::Dying])
 		return;
 
-	//Repair the hull...or maybe unrepair?
-	if(sip->hull_repair_rate != 0.0f)
+	// Repair the hull...or maybe unrepair?
+	if (!fl_equal(sip->hull_repair_rate, 0.0f))
 	{
-		objp->hull_strength += sp->ship_max_hull_strength * sip->hull_repair_rate * frametime;
+		float max_hull_repair_strength = sp->ship_max_hull_strength * sip->hull_repair_max;
+		objp->hull_strength += max_hull_repair_strength * sip->hull_repair_rate * frametime;
 
-		if (objp->hull_strength > sp->ship_max_hull_strength)
-			objp->hull_strength = sp->ship_max_hull_strength;
+		if (objp->hull_strength > max_hull_repair_strength)
+			objp->hull_strength = max_hull_repair_strength;
```

Comment:
This is admittedly unlikely, but this will make the ship unable to regain a hull strength above its hull repair max, regardless of the source. For instance, while the hull repair is active, and a SEXP causes it to repair above the threshold, this will bring it back down again. Instead of capping the `objp->hull_strength` the added health due to hull repair should itself be limited to only adding if below the threshold.

Reply:
Changes committed to fix this following similar method that the guardian code uses (per Discord discussion).